Refactor formatter to use regex

Davi-S opened the pull request:
This pull request completely refactors the `formatString` logic, improving reliability and maintainability.

The previous implementation, which relied on manual index tracking, has been replaced with a regex-based approach.

*Comparison:*
- Improved readability
- Robust escape handling
- Less specific error location

Davi-S committed:
New refactor codebase

This new function uses regex to replace the templates. It has a more clean and simple logic

Davi-S committed:
Separate code into smaller functions

Davi-S committed:
Add error handling and comments

Samsu-F commented:
To be completely honest, I do not think this is a good change. You list "Improved readability" and "improving reliability and maintainability" as a reason, but I'd strongly disagree. In my opinion, this new version is harder to understand, and as a result, harder to maintain and more error-prone.

Just to drive the point home:
With the default config, you get an error if you enter a '{' into the username field. After completing it to "{username}", I just had to restart my PC because my whole system stopped responding.

Davi-S commented:
If so, then it really doesn't pay. I just tried to play a little with Regex, but apparently it wasn't as good as the original.

Davi-S commented:
> With the default config, you get an error if you enter a '{' into the username field. After completing it to "{username}", I just had to restart my PC because my whole system stopped responding.

Same here. This is scary lol. I'm sure this bug could be fixed, but it is not worth it

Davi-S commented:
I just tried to play a little with Regex, but apparently it wasn't as good as the original.

> With the default config, you get an error if you enter a '{' into the username field. After completing it to "{username}", I just had to restart my PC because my whole system stopped responding.

This is scary lol. I'm sure this bug could be fixed, though

> In my opinion, this new version is harder to understand, and as a result, harder to maintain and more error-prone.

I disagree with it. But the reality is that we are each prone to understanding our own codes better (aka, we are biased).

One empirical point is that the current code lacks comments.

Davi-S committed:
Fix error when using especial characters on the placeholder values

The bug was that you get an error if you enter a '{' into the username field. And after completing it to "{username}", the program just crashes due to infinity loop

Davi-S committed:
Reorder functions

Samsu-F commented:
The new implementation still relies on the fact that the escape character is '%'. The template `"%{%?%:%%%}"` works fine, but if you change `escapeCharacter` to `'#'` and use the template `"#{#?#:###}"`, it breaks.

Davi-S committed:
Fix cross-placeholder reference bug in conditional evaluation

When evaluating nested templates, the conditional evaluation was incorrectly treating resolved
placeholder values as new placeholder keys. This caused unintended cross-
references between different placeholders

Fixed ensuring that the conditional expressions evaluate based on the actual
placeholder values rather than creating recursive placeholder

Davi-S committed:
Replace hard coded values with the escapeCharacter property
